### PR TITLE
Comprehensive mvi integration specs and improvements

### PR DESCRIPTION
**Pull request by saneshark**
## Description of change
This allows for using a Thread local variable for storing an initialized version of the MVI class for the entirety of the request, such that single request does not make repeated external requests (either to Redis or external service). NOTE: for an internal server error or gateway timeout we do not cache to redis, and it was discovered that we can make up to 12 external requests in a single thread when that happens.

Unfortunately Rspec does not clear Thread local variables between runs, so I've had to also add that in which makes somewhat difficult to test the safety of threads. This is something I'm a little nervous about, but I guess you have to trust the Ruby and Rails core people. One possible fix might be to use something like `Thread.current[user.uuid] = { mvi: Mvi.new }` but that ends up making things a lot more complicated to destroy in specs.

Another thought was to use `ObjectSpace.find(<mvi_original.object_id>)` but that could run into race conditions with GC.

I also like this approach and would be open to that as well, since Rack is the perfect place seeing as we're concerned with the entire lifecycle of a request: https://github.com/steveklabnik/request_store

## Testing done
Just specs so far.

## Testing planned
We can really only reliably see this in action when it is in production. But we can ensure that there are no side-effects in staging.

#### Unique to this PR

- [ ] _Replace this line with individual tasks unique to your PR_

#### Applies to all PRs

- [ ] Appropriate logging
- [ ] Swagger docs have been updated, if applicable
- [ ] Provide link to originating GitHub issue, or connected to it via ZenHub
- [ ] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

**Comment by dhall** [comment]
I get about 250 failing specs locally are you seeing this behavior as well @saneshark

**Comment by saneshark** [comment]
@dhall yes! I've got some fixes in for some of them, mostly around the stub_mvi method. But encountered another issue with Marshall not being able to dump a Singleton class. Turns out Mvi object has a reference to user, and user has a reference to MVI, and user also has references to other services some of them have the Singleton configuration class... I'm starting to think maybe a JSON dump / load is better, otherwise i might create a User struct that doesn't have reference to MVI.

**Comment by jkassemi** [comment]
I think you're right to be concerned about the risks of the thread local approach; to effectively and safely take this approach generally you need to guarantee that a request that will be accessing the thread local initially resets that thread local. I've usually done that with a `before_action` in the application controller. 

Thread locals are globals. I've got no practical objection to globals, but they are sometimes (and I think this case qualifies), a solution around a less than perfect implementation. I think it's best to abstract their scope as much as possible - using them to define the `current_user` isn't so bad. 

In this case, if the thread local is the best approach, it would be a good idea to document known alternatives to make the argument. Is there an effective alternative to force access of the mvi profile through the `user` object, and memoize the mvi profile there? What are the actual ramifications of just caching the request to MVI itself instead of making the update here? Is the introduction of this smell preferable to lower level solutions (caching at the HTTP layer) or interface updates (the memoization from earlier)? Is it _required_ for performance purposes?

**Comment by saneshark** [comment]
I think I need to do a little bit more research on this. I'm going to see if I could use a Rack middleware instead so that I can rely on my tests a little bit better seeing as Rspec does not handle threads very well. 

I'm also going to research what it would take to refactor MVI in such a way that this is no longer an issue, but the problem there is difficult, seeing as our specs are written in a somewhat fragile manner in how they stub external requests to MVI right now.

**Comment by saneshark** [comment]
OK, so this fixes all of the places that were reloading user using User.find

session.rb
mhv_account.rb (multiple times)

And that appears to has to have gotten the specs around user_request_spec to pass.

I still think it's very convenient to be able to do a User.find(some_uuid), and I would like the expectation to be that it fetches from in memory before attempting to fetch from Redis, and before attempting to fetch from an external repository. The code smell is still there to present itself again in the future.

In any case, this ensures users controller latency doesnt go bananas.

File: app/models/mvi.rb [hunk 1]
```diff
@@ -26,92 +26,68 @@ def self.for_user(user)
...
-  def status
-    return MVI::Responses::FindProfileResponse::RESPONSE_STATUS[:not_authorized] unless @user.loa3?
-    mvi_response.status
-  end
-
   # A DOD EDIPI (Electronic Data Interchange Personal Identifier) MVI correlation ID
   # or nil for users < LOA 3
   #
   # @return [String] the edipi correlation id
-  def edipi
```
Comment (annaswims) [inline app/models/mvi.rb]: I like this pattern.

**Comment by saneshark** [comment]
@patrickvinograd if the specs are passing for user_request_spec, it just means /users wont make more than one external request, so you're right to be concerned that other places might be affected, but I don't think its necessary to block merging. We know that it's possible that there are other places we will likely call out more than once. In particular if as worker that depends on MVI data runs into failures, it will retry repeatedly which is a problem in and of itself. I'm not trying to fix all of the places where User.find could be a problem, just the places that affect identity

**Comment by saneshark** [comment]
in a future PR we might look at a more holistic approach to the problem

**Comment by patrickvinograd** [comment]
Ok, I'm on board with this, just wanted to understand what was changing.

**Pull request merged**